### ISSUE TITLE
Sporadic results in XCTest

### ISSUE THREAD

**Issue by bredfield** [realm/realm-cocoa#1350]
Using Xcode 6.1.1, Swift, and Realm 0.88, I'm getting some weird results with testing RLMObjects.

When testing the success of a model's creation, tests will (in a seemingly random fashion) pass or fail. I'm assuming this relates to the race conditions, as described here: https://groups.google.com/d/msg/realm-cocoa/LZQou7g8or8/lSmoy5beTRIJ. 

I'm using the SwiftTestCase as a base test class: https://groups.google.com/d/msg/realm-cocoa/LZQou7g8or8/lSmoy5beTRIJ. With one addition: since my models always use the `defaultRealmPath`, I set the path to the test path in `setUp`: 

```
    override func setUp() {
        super.setUp()

        deleteRealmFilesAtPath(defaultRealmPath())
        deleteRealmFilesAtPath(testRealmPath())
        RLMRealm.setDefaultRealmPath(realmWithTestPath().path)
    }
```

When testing a model function that creates a user, something along the lines of this:

```
       ...
       let realm = RLMRealm.defaultRealm()
       realm.beginWriteTransaction()
       realm.addOrUpdateObject(user)
       realm.commitWriteTransaction()  
       return user
```

With a simple test such as:
```
        TestUser.createUserWithToken(token, email: testEmail)

        let predicate = NSPredicate(format: "token = %@", token)
        if let user = DCUser.objectsWithPredicate(predicate).firstObject() as? TestUser {
            XCTAssert(user.token == token, "User created with Token")
            XCTAssert(user.email == testEmail, "User created with email")
        }
        else {
            XCTFail("User didn't exist")
        }
```

The test will sporadically pass or fail.

Is there something else I should be doing to alleviate the issue?

**Comment by tgoyne** [comment]
How is it failing? Is the object not being found, or is there stale data from previous tests resulting in the asserts failing?

**Comment by bredfield** [comment]
Ah, I should have specified -- it always fails with the object being nil.

**Comment by tgoyne** [comment]
Are there any uses of in-memory realms or other places where `setDefaultRealmPath` is called?

**Comment by bredfield** [comment]
No in memory realms. I only set the default realm to the test one, in `setUp()`.

**Comment by bredfield** [comment]
I just tried doing the same without using the defaultRealm, but instead passing through the test realm to the createUser method, with the same sporadic result.

**Comment by tgoyne** [comment]
Are any of your tests doing thing on multiple threads? If there's a background thing using Realm that sometimes finishes and is deallocated after `setUp` is called for the next test, that could cause issues similar to what you're seeing.

**Comment by bredfield** [comment]
Nope, everything is on a single thread. Also, the above test example will fail even when run alone.

**Comment by jpsim** [comment]
Hi @bredfield, I'm sorry this is causing you some problems. It doesn't look like you're doing anything wrong from what you've told us.

We'll need access to an Xcode project where we can reproduce this issue to troubleshoot further. Do you think you could send us anything? You can email us code privately at help@realm.io if you prefer.

**Comment by bredfield** [comment]
Hey JP, thanks for looking into this. It's quite possible it is on my end. I've set up an example project that's exhibiting the symptoms: https://github.com/bredfield/RealmTest.

**Comment by tgoyne** [comment]
The problem is the `as?` -- `firstObject()` is returning a non-nil object of the correct type, but for some reason the dynamic cast is failing.

**Comment by bredfield** [comment]
Great find! Have you run into any similar dynamic casting issues previously? Is there a preferred way to cast to the object's class?

**Comment by dave-trudes** [comment]
+1

**Comment by omaxmedia90** [comment]
+1

**Comment by AnthonyMDev** [comment]
+1

**Comment by jpsim** [comment]
Hi all, thanks for your patience on this. The problem demonstrated in the example project created by @bredfield was caused by compiling an RLMObject subclass in both the app and test targets.

**tl;dr; don't add your app's Swift files to your unit test targets.**

Please read https://github.com/bredfield/RealmTest/pull/1 for a more detailed explanation of what was happening and how it can be fixed.
